File: 217-contains-duplicate/contains-duplicate.cpp
class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        // unordered_set<int> visited;
        // for (const auto& num: nums) {
        //     if (visited.count(num)) {
        //         return true;
        //     }
        //     visited.insert(num);
        // }
        // return false;

        unordered_set<int> s;

        for (int i = 0; i < nums.size(); i++) {
            if (s.find(nums[i]) != s.end()) return true;

            s.insert(nums[i]);
        }

        return false;

        // sort(nums.begin(), nums.end());
        // for (int i = 1; i < nums.size(); i++) {
        //     if (nums[i] == nums[i - 1]) {
        //         return true;
        //     }
        // }
        // return false;
    }
};

File: 217-contains-duplicate/contains-duplicate.py
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        # T: O(n), S: O(n)
        visited = set()
        for num in nums:
            if num in visited:
                return True
            visited.add(num)
        return False
        
        # # O(n), O(n) solution
        # visited = set()
        # for n in nums:
        #     if n in visited:
        #         return True
        #     visited.add(n)
        # return False

        # # O(nlogn), O(1) solution
        # nums.sort()
        # for i in range(1, len(nums)):
        #     if nums[i] == nums[i - 1]:
        #         return True
        # return False